File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-c1-client-host-lifecycle-scenario.ts
// RL-C1: 다량 connection 뒤 정상 종료한다 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { profileReq } from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';

export async function runRlC1(options: ClientOptions): Promise<void> {
  for (let index = 0; index < 12; index += 1) {
    const reply = await postJson<ProfileRes>(
      options.consumerUrl,
      '/profile/request/new-client',
      profileReq(`rl-c1-${index}`)
    );
    ensure(reply.value === 'profile:fast', 'RL-C1 request failed before cleanup.');
  }

  const followUp = await postJson<ProfileRes>(
    options.consumerUrl,
    '/profile/request/new-client',
    profileReq('rl-c1-after-cleanup')
  );
  ensure(followUp.value === 'profile:fast', 'RL-C1 follow-up failed after client cleanup.');

  const lifecycleEvidence = await Promise.any([
    postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: 'marker=rl-c1-' }),
    postJson<string[]>(options.providerBUrl, '/evidence/wait', { contains: 'marker=rl-c1-' })
  ]);
  ensure(
    lifecycleEvidence.some((line) => line.includes('marker=rl-c1-')),
    'RL-C1 did not record expected lifecycle evidence.'
  );

  const followUpEvidence = await Promise.any([
    postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: 'marker=rl-c1-after-cleanup' }),
    postJson<string[]>(options.providerBUrl, '/evidence/wait', { contains: 'marker=rl-c1-after-cleanup' })
  ]);
  ensure(
    followUpEvidence.some((line) => line.includes('marker=rl-c1-after-cleanup')),
    'RL-C1 did not record expected cleanup follow-up evidence.'
  );

  console.log('scenario RL-C1 passed');
}

File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlC1ClientHostLifecycleScenario.cs
// Verifies RL-C1 Client Host Lifecycle behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-C1 verifies repeated client lifecycle traffic and a follow-up request.
internal static class RlC1ClientHostLifecycleScenario
{
    public static async Task RunAsync(
        ClientOptions options,
        ZLinkHttpClient consumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        foreach (var index in Enumerable.Range(0, 12))
        {
            var reply = await EphemeralRouteClient.RequestAsync(
                options,
                new ProfileReq("fast", $"rl-c1-{index}"));
            ZlinkStreamAssert.Ensure(
                reply.Value == "profile:fast",
                "RL-C1 request failed before cleanup.");
        }

        var followUp = await EphemeralRouteClient.RequestAsync(
            options,
            new ProfileReq("fast", "rl-c1-after-cleanup"));
        ZlinkStreamAssert.Ensure(followUp.Value == "profile:fast", "RL-C1 follow-up failed after client cleanup.");

        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-c1-"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(new EvidenceWaitReq(["marker=rl-c1-"], []))
                .Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(evidence.Any(line => line.Contains("marker=rl-c1-", StringComparison.Ordinal)),
                "RL-C1 did not record expected evidence 'marker=rl-c1-'.");
        }
        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait")
                .Body(new EvidenceWaitReq(["marker=rl-c1-after-cleanup"], [])).Async<string[]>(timeout.Token)
                .AsTask();
            var waitB = providerB.Post("/evidence/wait")
                .Body(new EvidenceWaitReq(["marker=rl-c1-after-cleanup"], [])).Async<string[]>(timeout.Token)
                .AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(
                evidence.Any(line => line.Contains("marker=rl-c1-after-cleanup", StringComparison.Ordinal)),
                "RL-C1 did not record expected evidence 'marker=rl-c1-after-cleanup'.");
        }

        Console.WriteLine("scenario RL-C1 passed");
    }
}
